lib/sprockets/sass.rb

```diff
@@ -1,29 +1,131 @@
+# frozen_string_literal: true
 require 'sprockets'
 require 'sprockets/sass/version'
+require 'sprockets/sass/utils'
 require 'sprockets/sass/sass_template'
 require 'sprockets/sass/scss_template'
-require 'sprockets/engines'
+require 'sprockets/sass/functions'
+require 'json'
 
+# the module of Sprockets
 module Sprockets
+  # The internal Sass module used to load and acessing configuration
   module Sass
+    autoload :LegacyCacheStore, 'sprockets/sass/legacy_cache_store'
     autoload :CacheStore, 'sprockets/sass/cache_store'
     autoload :Compressor, 'sprockets/sass/compressor'
     autoload :Importer,   'sprockets/sass/importer'
-    
     class << self
       # Global configuration for `Sass::Engine` instances.
       attr_accessor :options
-      
+
       # When false, the asset path helpers provided by
       # sprockets-helpers will not be added as Sass functions.
       # `true` by default.
       attr_accessor :add_sass_functions
+
+      def version_of_sprockets
+        Sprockets::VERSION.split('.')[0].to_i
+      end
     end
-    
+
     @options = {}
     @add_sass_functions = true
   end
-  
-  register_engine '.sass', Sass::SassTemplate
-  register_engine '.scss', Sass::ScssTemplate
+
+  begin
+    require 'sprockets/directive_processor'
+    require 'sprockets/sass_processor'
+    require 'sprockets/sassc_processor'
+    require 'sprockets/digest_utils'
+    require 'sprockets/engines'
+  rescue LoadError; end
+
+  if Sprockets::Sass.version_of_sprockets >= 3
```

Comment:
Maybe since there's so much logic associated with registration now, this could be broken down into a methods called conditionally based on the version